### DISCUSSION TITLE
Change destination/redirect to another URL after publish

### DISCUSSION THREAD

**Discussion by PiemP** [OrchardCMS/OrchardCore#13910] [Q&A] [answered]
Good evening, 

I would like to change/redirect user after publish a content item but I don't know how to do it. I have tried to create a specific content handler class that implement `ContentHandlerBase` and the method `GetContentItemAspectAsync` to change the value of `AdminRouteValues` but I believe to do something wrong with the excecution order of the handlers. 

Is it the right way to do this things? 

Thank you.

**Comment by MikeAlhayek** [comment]
`GetContentItemAspectAsync ` would allow you to change the URL for a link like `Display` or `Edit` link but not the ReturnUrl for the Edit form.

you should be able to change the `Context.Request.Query["ReturnUrl"]` using an action filter which will change the return URL in the Edit.cshtml view.

You should look at the code in `OrchardCore.Contents\Admin\Edit.cshtml`

**Reply by PiemP** [reply]
Thank you @MikeAlhayek for your answer but the problem is not with the `Edit` operation but is with `Create` operation. When I create a content item I don't have the `contentItemId` and I don't know how to get it and use it in an `ActionFilterAttribute`. In the Confirm view I wanna show some values taked from the content item published.

With a `ContentHandlerBase` is more easy to check the type of the content item and redirect the user to a Confirm view only in some cases. But seem my `ContentHandlerBase` is called before the other ones and I can't change the value of `AdminRouteValues `.

**Reply by MikeAlhayek** [reply, marked as answer]
You can use `ContentHandlerBase` to add the new created ContentItem to the HttpContext.Feature collection and then use a filter to access that new contenItem with something like this

```
public class CustomContentHandler : ContentHandlerBase
{
    public override Task CreatedAsync(CreateContentContext context)
    {
        if(context.ContentItem.ContentType == "CustomType")
        {
        	// either this
        	_httpContentAccessssor.HttpContext.Request.Query["ReturnUrl"] = "";

        	// or do this and use an action filter to read it and change the ReturnUtr later.
        	_httpContentAccessssor.HttpContext.Features.Set(new NewContentItemFeature(context.ContentItem))

        }
    }
}
```

**Reply by MikeAlhayek** [reply]
@PiemP If you want to use `ContentItemMetadata`, you'll need to submit a PR changing https://github.com/OrchardCMS/OrchardCore/blob/7cb1cd2dc69ff5c5305e9feee80ca59d43d793b4/src/OrchardCore.Modules/OrchardCore.Contents/Controllers/AdminController.cs#L643 to use `TryAdd` instead of `Add`. Then you should be able to use `GetContentItemAspectAsync`

**Reply by PiemP** [reply]
Thank you very much @MikeAlhayek,

You have saved me. I did not know `Feature` and `Items` in the `HttpContext`. It's extremely useful.

About the PR I can make it but the solution with the filter can overwrite `returnUrl` in all cases. The `TryAdd` solution works only if the user press save/publish and continue. For this I don't know if a PR about the change from `Add` to `TryAdd` could be really useful.

With a filter it's possible to make a "simple" overwrite of the `returnUrl` in this way:  

```
...

public class CustomFilter : ActionFilterAttribute
    {
        private readonly ILogger<PraticheConfirmFilter> _logger;
        private readonly AdminOptions _adminOptions;
        private readonly LinkGenerator _generator;


        public CustomFilter(ILogger<CustomFilter> logger, IOptions<AdminOptions> adminOptions, LinkGenerator generator)
        {
            _logger = logger;
            _adminOptions = adminOptions.Value;
            _generator = generator;
        }

        
        public override async void OnActionExecuting(ActionExecutingContext filterContext)
        {
            if (filterContext == null)
            {
                throw new ArgumentNullException(nameof(filterContext));
            }

            if (!await CanBeUsed(filterContext.HttpContext))
                return;

            
           //this is an example and works only in edit mode, see contentItemId taked from ActionArgument
            filterContext.ActionArguments["returnUrl"] = _generator.GetPathByName(filterContext.HttpContext, "ConfirmView", new { area = "CustomAre", contentItemId = filterContext.ActionArguments["contentItemId"] }) ?? filterContext.ActionArguments["returnUrl"];
        }
       ... 
``` 

My final solution is implemented in the `OnActionExecuted` method: 

```
     public override async void OnActionExecuted(ActionExecutedContext filterContext)
        {
            if (filterContext == null)
            {
                throw new ArgumentNullException(nameof(filterContext));
            }

            if (!await CanBeUsed(filterContext.HttpContext))
                return;


            //if is a CTYPE and only if I publish it
            var st = filterContext.HttpContext?.Features?.Get<StereotypeContentItemFeature>() ?? new StereotypeContentItemFeature();
            if (Stereotype.CTYPE.Equals(st.Stereotype, StringComparison.InvariantCultureIgnoreCase) && st.IsAPublish)
            {
                //if is a create get contentitemid from features, if is an edit get it from route data
                var ci = filterContext.HttpContext?.Features?.Get<NewContentItemFeature>()?.ContentItem?.ContentItemId ?? filterContext.RouteData.Values["contentItemId"];
                var returnUrl = _generator.GetPathByName(filterContext.HttpContext, "ConfirmView", new { area = "CustomArea", contentItemId = ci });
                filterContext.Result = ((Controller)filterContext.Controller).LocalRedirect(returnUrl, true);
            }
        }
```

What I try to do before with my custom `ContentHandlerBase` is: 

```
    public override Task GetContentItemAspectAsync(ContentItemAspectContext context)
        {

            var contentTypeDefinition = _contentDefinitionManager.GetTypeDefinition(context.ContentItem.ContentType);
            //check is a pratica
            var st = contentTypeDefinition.GetStereotype();
            if (!Stereotype.CTYPE.Equals(st, StringComparison.InvariantCultureIgnoreCase))
            {
                _logger.LogInformation($"is not a pratica: `{st}`");
                return Task.CompletedTask;
            }

            _httpContextAccessor.HttpContext?.Features?.Set(new StereotypeContentItemFeature(st));

            
            //TODO: check I'm doing a publish action

            return context.ForAsync<ContentItemMetadata>(metadata =>
            {
                //change AdminRouteValues to show confirm page after publish, rember to use publish and Continue button
                //otherwise returnUrl rules
                metadata.AdminRouteValues = new RouteValueDictionary {
                        {"Area", "CustomArea"},
                        {"Controller", "Admin"},
                        {"Action", "Confirm"},
                        {"Name", "ConfirmView"},
                        {"ContentItemId", context.ContentItem.ContentItemId}
                    };


                return Task.CompletedTask;
            });
        
``` 
But I can't make it work.
